src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -309,6 +312,59 @@ TransactionManager::submit_transaction_direct(
     trim_alloc_to);
 }
 
+TransactionManager::update_lba_mappings_ret
+TransactionManager::update_lba_mappings(
+  Transaction &t,
+  std::list<CachedExtentRef> &pre_allocated_extents)
+{
+  LOG_PREFIX(TransactionManager::update_lba_mappings);
+  SUBTRACET(seastore_t, "update extent lba mappings", t);
+  return seastar::do_with(
+    std::list<LogicalCachedExtentRef>(),
+    std::list<CachedExtentRef>(),
+    [this, &t, &pre_allocated_extents](auto &lextents, auto &pextents) {
+    auto chksum_func = [&lextents, &pextents](auto &extent) {
+      if (!extent->is_valid() ||
+          !extent->is_fully_loaded() ||
+          // EXIST_MUTATION_PENDING extents' crc will be calculated when
+          // preparing records
+          extent->is_exist_mutation_pending()) {
+        return;
+      }
+      if (extent->is_logical()) {
+        // for rewritten extents, last_committed_crc should have been set
+        // because the crc of the original extent may be reused.
+        // also see rewrite_logical_extent()
+        if (!extent->get_last_committed_crc()) {
+          extent->update_checksum();
+        }
+        assert(extent->calc_crc32c() == extent->get_last_committed_crc());
+        lextents.emplace_back(extent->template cast<LogicalCachedExtent>());
+      } else {
+        pextents.emplace_back(extent);
+      }
+    };
+
+    t.for_each_finalized_fresh_block(chksum_func);
+    t.for_each_existing_block(chksum_func);
+    std::for_each(
+      pre_allocated_extents.begin(),
+      pre_allocated_extents.end(),
+      chksum_func);
```

Comment:
Let's remember the dependencies here.

If my understandings are correct:
```c++
// The order of ool-writes and lba-mapping updates:
// 1. Write delayed-ool fresh logical extents.
// 2. For delayed-ool fresh logical extents, update lba-leaf crc and paddr.
// 2. For other fresh logical extents, update lba-leaf crc.
t.for_each_finalized_fresh_block(chksum_func);
// 2. For existing-clean logical extents, update lba-leaf crc.
t.for_each_existing_block(chksum_func);
// 2. For pre-allocated fresh logical extents, update lba-leaf crc.
// 2. For inplace-rewrite dirty logical extents, update lba-leaf crc.
std::for_each(
  pre_allocated_extents.begin(),
  pre_allocated_extents.end(),
  chksum_func);
// 3. For fresh physical extents, update in-extent crc.
// 3. For pre-allocated fresh physical extents, update in-extent crc.
// 4. Write pre-allocated logical and physical extents.
```

Reply:
Ah, will add comments